upickle/shared/src/main/scala/upickle/Types.scala

```diff
@@ -90,9 +90,31 @@ trait Types{ types =>
   trait Reader[T]{
     def read0: PF[Js.Value, T]
 
-    final def read : PF[Js.Value, T] = ({
+    private def readNull: PF[Js.Value, T] = {
       case Js.Null => null.asInstanceOf[T]
-    }: PF[Js.Value, T]) orElse read0
+    }
+
+    private def invalid: PF[Js.Value, T] = {
+      case any => throw Invalid.Data(any, "name")
+    }
+
+    /*
```

Comment:
This huge comment block should be in the pull request text, not embedded in the codebase